target/linux/ath79/dts/ar9344_tplink_cpe510-v1.dts

```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "ar9344.dtsi"
+
+/ {
+	model = "TP-Link CPE510 v1";
+	compatible = "tplink,cpe510-v1", "qca,ar9344";
+
+	aliases {
+		led-boot = &system;
+		led-failsafe = &system;
+		led-running = &system;
+		led-upgrade = &system;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		lan0 {
+			label = "tp-link:green:lan0";
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+		};
+
+		lan1 {
+			label = "tp-link:green:lan1";
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
+		};
+
+		link1 {
+			label = "tp-link:green:link1";
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+
+		link2 {
+			label = "tp-link:green:link2";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+
+		link3 {
+			label = "tp-link:green:link3";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+
+		system: link4 {
+			label = "tp-link:green:link4";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&jtag_disable_pins>;
+
+		reset {
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+	};
+
+	gpio-export {
+		compatible = "gpio-export";
+
+		gpio_ext_lna0 {
+			gpio-export,name = "tp-link:ext:lna0";
+			gpio-export,output = <1>;
+			gpios = <&gpio 18 GPIO_ACTIVE_HIGH>;
+		};
+
+		gpio_ext_lna1 {
+			gpio-export,name = "tp-link:ext:lna1";
+			gpio-export,output = <1>;
+			gpios = <&gpio 19 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&ref {
+	clock-frequency = <40000000>;
+};
+
+&uart {
+	status = "okay";
+};
+
+&gpio {
+	status = "okay";
+};
+
+&spi {
+	num-cs = <1>;
+
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			uboot: partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x020000>;
+				read-only;
+			};
+
+			partition@20000 {
+				label = "partition-table";
+				reg = <0x020000 0x010000>;
+				read-only;
+			};
+
+			productinfo: partition@30000 {
+				label = "product-info";
+				reg = <0x030000 0x010000>;
+				read-only;
+			};
+
+			partition@40000 {
```

Comment:
>  I'm not sure if target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_tplink.c works for tplink-safeloader. If it works a compatible string of "tplink,firmware" here should do the job.

That was my first approach (compatible="tplink,fimware"). This resulted in a crash as "unable to mount root filesystem" or similar. It did not detect the partitions.

Reply:
A crash **or** `unable to mount root filesystem`?

Please provide mtd part of the bootlog with a working kernel, so we can see how partitioning works exactly.

Reply:
Yeah, it was not a crash but unable to mount root and therefore reboot in one second.

This is (part of) the bootlog of the same device on ar71xx:

```
[    0.586756] m25p80 spi0.0: found s25fl064k, expected m25p80
[    0.604506] m25p80 spi0.0: s25fl064k (8192 Kbytes)
[    0.609423] 8 cmdlinepart partitions found on MTD device spi0.0
[    0.615452] Creating 8 MTD partitions on "spi0.0":
[    0.620319] 0x000000000000-0x000000020000 : "u-boot"
[    0.627656] 0x000000020000-0x000000030000 : "partition-table"
[    0.635982] 0x000000030000-0x000000040000 : "product-info"
[    0.643945] 0x000000040000-0x000000200000 : "kernel"
[    0.651458] 0x000000200000-0x0000007c0000 : "rootfs"
[    0.658473] mtd: device 4 (rootfs) set to be root filesystem
[    0.664273] 1 squashfs-split partitions found on MTD device rootfs
[    0.670614] 0x000000430000-0x0000007c0000 : "rootfs_data"
[    0.678682] 0x0000007c0000-0x0000007f0000 : "config"
[    0.686265] 0x0000007f0000-0x000000800000 : "ART"
[    0.693450] 0x000000040000-0x0000007c0000 : "firmware"
```

Reply:
OK, so it seems `ar71xx` was using a static `kernel` and `rootfs` partitions. If we indeed want to keep it that way then `fixed-partitions` is a right choice.

Reply:
So `firmware` should be defined as `reg = <0x40000 0x780000>`, `kernel` as `reg = <0x40000 0x1c0000>`, `rootfs` as `reg = <0x200000 0x5c0000>`

Reply:
@ynezz As the partitions `kernel` and `rootfs` are nested (in `firmware`) this is also represented by the DTS (and therefore, inside `firmware`, addresses are relative to it)
DTS would not allow two partitions at the same address anyway.

This is part of the ath79 bootlog with the DTS in review: (the device is operational)
```
[    0.370814] m25p80 spi0.0: s25fl064k (8192 Kbytes)
[    0.375782] 6 fixed-partitions partitions found on MTD device spi0.0
[    0.382225] Creating 6 MTD partitions on "spi0.0":
[    0.387122] 0x000000000000-0x000000020000 : "u-boot"
[    0.393127] 0x000000020000-0x000000030000 : "partition-table"
[    0.399916] 0x000000030000-0x000000040000 : "product-info"
[    0.406415] 0x000000040000-0x0000007c0000 : "firmware"
[    0.412467] 2 fixed-partitions partitions found on MTD device firmware
[    0.419161] Creating 2 MTD partitions on "firmware":
[    0.424218] 0x000000000000-0x0000001c0000 : "kernel"
[    0.430157] 0x0000001c0000-0x000000780000 : "rootfs"
[    0.436086] mtd: device 5 (rootfs) set to be root filesystem
[    0.441881] 1 squashfs-split partitions found on MTD device rootfs
[    0.448224] 0x0000003c0000-0x000000780000 : "rootfs_data"
[    0.454613] 0x0000007c0000-0x0000007f0000 : "config"
[    0.460529] 0x0000007f0000-0x000000800000 : "art"
```

Reply:
@ynezz: no. Offsets for the nested partitions are relative to the parent. So you need `0x40000 - 0x40000` and `0x200000 - 0x40000`. That is what @p-wassi did.

Reply:
> `fixed-partitions` is a right choice

@rmilecki Why? If mtd splitter for `tplink,fimware` doesn't work for this images, then it needs to be fixed. To me it seems like it's first device of this type (using some kind of kernel loader) being ported over to ath79 and that's maybe why it doesn't work out of the box with mtd splitter and needs to be fixed. Or am I mistaken and `fixed-partitions` is really the only and right choice?

Reply:
@p-wassi It would be really nice to know why you can't simply use just `firmware` with `reg = <0x40000 0x780000>` and `tplink,fimware` mtd splitter. Could you please add some `pr_info()` into `mtdsplit_parse_tplink()` and find out why does it fails, why it can't parse the flashed image? It's because it can't find/parse the image header or something else?